src/main/java/com/example/b2b_opportunities/Service/AuthenticationService.java

```diff
@@ -91,4 +99,51 @@ private boolean arePasswordsMatching(UserRequestDto userRequestDto) {
         return userRequestDto.getPassword().equals(userRequestDto.getRepeatedPassword());
     }
 
+    public LoginResponse oAuthLogin(Principal user) {
+        if (user instanceof OAuth2AuthenticationToken authToken) {
+            OAuth2User oauth2User = authToken.getPrincipal();
+            Map<String, Object> attributes = oauth2User.getAttributes();
+
+            String email = (String) attributes.get("email");
+
+            if (!isEmailInDB(email)) {
+                createUserFromOAuth(attributes);
+            }
+
+            return generateLoginResponse(email);
+        }
+        throw new ServerErrorException("Authentication failed: The provided authentication is not an OAuth2 token.");
+    }
+
+    private void createUserFromOAuth(Map<String, Object> attributes) {
+        RoleType roleUser = RoleType.ROLE_USER;
+        Role role = Role.builder()
+                .id(roleUser.getId())
+                .name(roleUser.name())
+                .build();
+
+        String email = (String) attributes.get("email");
+        User user = User.builder()
+                .username(email)
+                .firstName((String) attributes.get("given_name"))
+                .lastName((String) attributes.get("family_name"))
+                .email(email)
+                .provider("google")
```

Comment:
It is better to extract the `google` value as a Java constant.

Reply:
True. Fixed now.